### ISSUE TITLE
Unclear documentation around gracefully stopping a server and if it should leave or not

### ISSUE THREAD

**Issue by nvx** [hashicorp/nomad#7943]
The page https://www.nomadproject.io/docs/install/production/nomad-agent/ under the heading "Stopping an Agent" mentions this:

```
An agent can be stopped in two ways: gracefully or forcefully. By default, any signal to an agent (interrupt, terminate, kill) will cause the agent to forcefully stop. Graceful termination can be configured by either setting leave_on_interrupt or leave_on_terminate to respond to the respective signals.
```

Further it mentions for servers that a graceful leave is a good idea:

```
It is especially important that a server node be allowed to leave gracefully so that there will be a minimal impact on availability as the server leaves the consensus peer set. 
```

However in https://www.nomadproject.io/docs/configuration/ the documentation for the leave_on_* flags has this:

```
This value should only be set to true on server agents if it is expected that a terminated server instance will never join the cluster again.
```

Which seems contradictory. Being that Nomad maintains a raft consensus it makes sense that gracefully stopping a node should be preferred but not all instances that you want to stop a server will it never join the cluster again. Tasks like Nomad version or host OS updates require stopping a server agent, and being a planned event it makes sense that it's as seamless to clients and workloads as possible, but the server will definitely be joining the cluster again afterwards.

I recently noticed occasional errors when one of the server nodes was unavailable (not even a leader in this instance) as the host OS was rebooting for kernel updates using the nomad CLI:

```
Error determining leaders: 1 error occurred:
        * Region "xxx": Unexpected response code: 500 (rpc error: failed to get conn: dial tcp x.x.x.x:4647: connect: connection refused)
```

Which is what lead me down this path as to what the proper process should be during such operations.

**Comment by tgross** [comment]
Hi @nvx! Thanks for opening this. I definitely agree that the documentation for this needs some work. 

The reason for the wording "will never join the cluster again" has to do with a few things:
- The `leave_after_*` operation is expensive in terms of extra raft log messages
- In raft v2, the IP address of the node was used as the name, which means that you'd want to leave in the case where you were replacing the node but keeping the IP address. This also impacted users who had to migrate from raft v2 to v3.

For running an in-place rolling update of a cluster as you've described here, the primary things to watch out for are:
- Maintain quorum during the upgrade, so if you're in a 3 node cluster you don't want to restart them all at the same time. (Likewise, if you were going to update via replacement in a cloud environment, you don't want to have 3 servers add 3 more replacements, and then remove the 3 old ones as you'll be less than the new quorum of `(6 / 2) +1` nodes).
- After each server node is restarted leave the cluster is given time to come to consensus before continuing to the next server. To be really sure, you can use `nomad agent-info` vs each server to verify they all have the same `raft.applied_index`.

I'm going to mark this issue as a documentation issue and we'll come back to this with some expanded docs in the near future.

**Comment by nvx** [comment]
Yeah definitely at least part of this is a documentation issue, although from what you've written as long as you maintain a quorum you shouldn't get errors from the Nomad CLI when a node goes down - but this isn't the behaviour I saw. I did get intermittent 500 errors until the node was detected as "failed" in the `nomad operator raft list-peers` output even for a non-leader that was stopped while still having a quorum of servers remaining.

Perhaps a way for a exiting server to signal to the other servers that it should be marked as failed immediately rather than having to wait for the other servers to detect it (which seems to take on the order of tens of seconds which is a little slower than desirable) but not doing a full server leave might be a better way of handling such cases.

**Comment by Kamilcuk** [comment]
Thanks for answering! There are 3 relevant settings: leave_on_interrupt, leave_on_terminate, rejoin_after_leave. What is the preferred setting for these 3? If I do not care about raft log messages size and have raft v3, is the reasoning for it not relevant?

For example, I am making a configuration upgrade to one of the servers. Bottom line, I will be restarting the Nomad server, because this is the way to refresh configuration. I think I would want it to leave nicely, and join nicely, I have a lot of time to do it, and most importantly I do not want any operation downtime. They way I am doing it now, is to `nomad operator raft remove-peer  ; systemctl restart nomad` in one line, and then I am monitoring `nomad operator api /v1/operator/autopilot/health`.

**Comment by marco-m-pix4d** [comment]
Hello @tgross any chance we could update the documentation around gracefully stopping a server and if it should leave or not ? Thank you!

**Comment by tgross** [comment]
Yeah, this should have been addressed. I've tagged this to get it onto the triage board.

**Comment by aimeeu** [comment]
Closing this ticket in prep to move docs to the unified docs repo. Internal Jira ticket: https://hashicorp.atlassian.net/browse/CE-1078
